### DISCUSSION TITLE
Upcoming new libxml2 dependency for iDynTree

### DISCUSSION THREAD

**Discussion by traversaro** [robotology/community#316] [Announcements]
In the next days, a [new shiny URDF parser](https://github.com/robotology/idyntree/pull/460) with [much better diagnostic errors](https://github.com/robotology/idyntree/pull/460#issuecomment-400845404) will land on iDynTree `devel` branch, contributed by @francesco-romano at [DeepMind](https://deepmind.com/). 

The new parser is  based on the MIT-licensed [libxml2 library](http://xmlsoft.org/): as  a consequence, libxml2 will become a new **required dependency** of iDynTree, so if you are using the *devel* branch of iDynTree, you will need to have `libxml2` installed in your system. Check the following for the detailed instructions on the recommended way of installing it. 

# Debian/Ubuntu 
~~~
sudo apt install libxml2-dev 
~~~

# macOS 
`libxml2` is already part of the library distributed with the developement tools on macOS.

# Windows 

If you are using `vcpkg` to manage your dependencies, you can install libxml2 via `vcpkg` : 
~~~
./vcpkg install --triplet x64-windows libxml2 
~~~
where `x64-windows` should be changed to `x86-windows` if you are targeting 32-bit binaries. 

If you typically install your dependencies using the [YARP and ICUB dependencies installers](http://wiki.icub.org/wiki/Downloads), we recommend you to use the new [robotology-additional-dependencies](https://github.com/robotology-playground/robotology-additional-dependencies/releases/tag/v0.1.0), that is a new installer of dependencies generated by vcpkg that contains `libxml2`. This installer is automatically accounting for environment variables, similarly to how the YARP and ICUB dependencies installers do. 

Please report any issue or problem regarding to this, thanks!

**Comment by francesco-romano** [comment]
@traversaro note that macOS already ships with `libxml2`. 
If brew is used, it should be specified in some way to CMake, otherwise I think by  default the system one get picked.

**Comment by traversaro** [comment]
@francesco-romano so do you think it make sense **not** to install libxml2 using homebrew to avoid confusion?

**Comment by francesco-romano** [comment]
It is a good question.

From my investigations I found the following: (macOS 10.13.6):
- version: 2.9.4
- library is in /usr/lib/libxml2
- headers in /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/usr/include/libxml2 but it might be that I did not symlinked the standard libraries in /usr (it is something you do with Xcode. Anyway, this implies that you need Xcode, but I think this requirement should be verified in all cases).

The question is: what version of libxml2 is shipped in previous OS versions?
I do not know the frequency of updates of the libraries in macOS -  note that 2.9.4 is 2 years old (https://gitlab.gnome.org/GNOME/libxml2/tags)

So we might want to setup some instructions to use brew, just to stay on the safe side.

**Comment by traversaro** [comment]
I checked and `libxml2` is present in most recent versions of Developer SDKs on macOS . I assume CMake has no trouble in finding it, right? For this reason, I updated the instructions relatively to `macOS` . 

By the way, the parser have been merged in the `devel` branch, see https://github.com/robotology/idyntree/pull/460 .

**Comment by traversaro** [comment]
The parser will be part of the upcoming iDynTree 0.11 .

**Comment by francesco-romano** [comment]
Sounds good to me. Let's keep it in this way. If in the future we realise we need a more recent version, we can switch to the one provided by homebrew
